File: code_samples/section9/problems/problem9_1/problem9_1.cs
/**
 * Computes the length of the Longest Increasing Subsequence (LIS) in `nums`.
 *
 * This uses the classic O(n log n) "tails" technique (often described via
 * patience sorting):
 *
 * Invariant:
 *   tails[i] is the smallest possible tail value of any strictly increasing
 *   subsequence of length (i + 1) seen so far.
 *
 * Why this works:
 *   Keeping tails as small as possible maximizes the chance to extend these
 *   subsequences with future numbers. `tails` remains sorted, enabling binary search.
 *
 * Update step for each element x:
 *   - Find the first index `lo` such that tails[lo] >= x (a lower_bound).
 *   - If lo == tails.Count: x extends the longest subsequence so far => append.
 *   - Else: replace tails[lo] with x to lower the tail for that subsequence length.
 *
 * Note:
 *   This returns only the LIS length, not the subsequence itself. Reconstructing
 *   the subsequence requires tracking predecessor indices.
 *
 * Complexity:
 *   Time:  O(n log n) (binary search per element)
 *   Space: O(n) (tails list)
 *
 * Edge cases:
 *   - Empty input => tails remains empty => returns 0.
 */
static int LengthOfLIS(int[] nums) {
    // `tails` holds the best (smallest) ending values for subsequences by length.
    var tails = new List<int>();

    // Process each number in the input array.
    foreach (int x in nums) {
        // Binary search for the first index where tails[index] >= x.
        // Search range is [lo, hi), i.e., hi is exclusive.
        int lo = 0, hi = tails.Count;

        while (lo < hi) {
            // Midpoint of the current search range.
            // Using (lo + hi) / 2 is safe here because indices are small.
            int mid = (lo + hi) / 2;

            // If tails[mid] < x, then x must go to the right of mid
            // to maintain sorted order (we're finding first >= x).
            if (tails[mid] < x) {
                lo = mid + 1;
            } else {
                // tails[mid] >= x, so mid is a candidate insertion point;
                // continue searching left to find the first such position.
                hi = mid;
            }
        }

        // After the loop, `lo` is the lower_bound position for x.
        if (lo == tails.Count) {
            // x is larger than all tails values => extends LIS length by 1.
            tails.Add(x);
        } else {
            // Replace existing tail with x to improve (lower) the tail value
            // for subsequences of length lo+1.
            tails[lo] = x;
        }
    }

    // The number of tails entries equals the length of the LIS.
    return tails.Count;
}

/**
 * Simple helper to run a test case and print:
 *  - test name
 *  - input array
 *  - computed LIS length
 *  - expected LIS length
 *
 * @param name     Descriptive test name/header.
 * @param arr      Input array for the LIS function.
 * @param expected Expected LIS length.
 */
static void Test(string name, int[] arr, int expected)
{
    // Run the algorithm under test.
    int result = LengthOfLIS(arr);

    // Print a small, readable report.
    Console.WriteLine(name);
    Console.WriteLine($"Input: [{string.Join(",", arr)}]");
    Console.WriteLine($"LengthOfLIS = {result} (expected {expected})\n");
}

// Header for the test run output.
Console.WriteLine("=== Testing LengthOfLIS ===\n");

// Test 1: Strictly increasing => LIS uses all elements.
Test("Test 1: Increasing sequence", [1, 2, 3, 4, 5], 5);

// Test 2: Classic mixed case => known LIS length is 4 (e.g., 2,3,7,101).
Test("Test 2: Mixed sequence", [10, 9, 2, 5, 3, 7, 101, 18], 4);

// Test 3: All identical => strictly increasing disallows duplicates => LIS = 1.
Test("Test 3: All identical", [7, 7, 7, 7, 7], 1);

// Test 4: Empty input => LIS = 0.
Test("Test 4: Empty array", Array.Empty<int>(), 0);

// Test 5: Strictly decreasing => best you can do is any single element => LIS = 1.
Test("Test 5: Decreasing sequence", [9, 7, 5, 3, 1], 1);

// Test 6: Random pattern => LIS length is 3 (e.g., 4,8,9).
Test("Test 6: Random pattern", [4, 10, 4, 3, 8, 9], 3);

File: code_samples/section9/problems/problem9_1/problem9_1.js
/**
 * Computes the length of the Longest Increasing Subsequence (LIS)
 * in the given array of numbers.
 *
 * This implementation uses the classic O(n log n) "tails" method,
 * sometimes described via patience sorting.
 *
 * Core invariant:
 *   tails[i] is the smallest possible tail value of any strictly
 *   increasing subsequence of length (i + 1) encountered so far.
 *
 * Because `tails` is always kept sorted, we can binary-search it
 * to decide where each new value belongs.
 *
 * Algorithm overview:
 *   For each value x in nums:
 *     1) Binary search tails to find the first index `lo` such that
 *        tails[lo] >= x (lower_bound).
 *     2) Place x at tails[lo].
 *        - If lo equals the current length of tails, x extends the LIS.
 *        - Otherwise, x replaces an existing tail, improving (lowering)
 *          the tail value for subsequences of that length.
 *
 * Important notes:
 *   - This function returns only the length of the LIS, not the subsequence.
 *   - The subsequence is strictly increasing (duplicates do not extend it).
 *
 * Complexity:
 *   Time:  O(n log n)
 *   Space: O(n)
 *
 * @param {number[]} nums - Input array of numbers
 * @returns {number} Length of the longest strictly increasing subsequence
 */
function lengthOfLIS(nums) {
    // `tails` holds candidate tail values for increasing subsequences.
    const tails = [];

    // Process each number in the input.
    for (const x of nums) {
        // Binary search in the range [0, tails.length)
        let lo = 0, hi = tails.length;

        while (lo < hi) {
            // Midpoint calculation using bit shift for fast integer division by 2.
            const mid = (lo + hi) >> 1;

            // If tails[mid] < x, x must go to the right of mid.
            if (tails[mid] < x) {
                lo = mid + 1;
            } else {
                // Otherwise, mid is a candidate insertion/replacement position.
                hi = mid;
            }
        }

        // Place x at index lo.
        // If lo == tails.length, this effectively appends x and increases LIS length.
        // Otherwise, it replaces an existing value to improve the tail.
        tails[lo] = x;
    }

    // The number of elements in tails equals the LIS length.
    return tails.length;
}

// ===========================
// Tests
// ===========================

/**
 * Helper to run a test case and print results in a consistent format.
 *
 * @param {string} name - Test name/description
 * @param {number[]} arr - Input array
 * @param {number} expected - Expected LIS length
 */
function test(name, arr, expected) {
    const result = lengthOfLIS(arr);
    console.log(`${name}`);
    console.log(`Input: [${arr.join(",")}]`);
    console.log(`lengthOfLIS = ${result} (expected ${expected})\n`);
}

console.log("=== Testing lengthOfLIS ===\n");

// Strictly increasing sequence => LIS uses all elements.
test("Test 1: Increasing sequence", [1, 2, 3, 4, 5], 5);

// Classic mixed example => known LIS length is 4.
test("Test 2: Mixed sequence", [10, 9, 2, 5, 3, 7, 101, 18], 4);

// All identical values => strictly increasing disallows duplicates => LIS = 1.
test("Test 3: All identical", [7, 7, 7, 7, 7], 1);

// Empty input => LIS length is 0.
test("Test 4: Empty array", [], 0);

// Strictly decreasing sequence => best LIS is any single element.
test("Test 5: Decreasing sequence", [9, 7, 5, 3, 1], 1);

// Random pattern => one LIS is [4,8,9], length = 3.
test("Test 6: Random pattern", [4, 10, 4, 3, 8, 9], 3);
